### PR TITLE
feat: implement e, pi, factorial, and hypot math functions

### PR DESCRIPTION
**Pull request by BABTUNA**
## Summary
- Adds 4 math functions from the missing list in #3793:
  - **`e()`** — Returns Euler's number (2.718...), zero-argument constant
  - **`pi()`** — Returns pi (3.14159...), zero-argument constant
  - **`factorial(expr)`** — Returns factorial of a non-negative integer, returns NaN for negative or non-integer inputs
  - **`hypot(a, b)`** — Returns sqrt(a² + b²), follows the existing `atan2` binary function pattern
- All 4 are registered in the Rust FUNCTION_REGISTRY (auto-available in SQL)
- Python wrappers and exports added

Resolves part of #3793

## Test plan
- [x] 8 new tests covering all 4 functions (constants, edge cases, bad input, None handling)
- [x] All 330 tests in `test_numeric.py` pass with no regressions
- [x] `ruff check` and `ruff format` clean
- [x] `cargo fmt` clean

### CODE REVIEW

File: src/daft-functions/src/numeric/factorial.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    if n.is_nan() || n < 0.0 || n != n.floor() {
+        return f64::NAN;
+    }
+    let n = n as u64;
+    let mut result: f64 = 1.0;
+    for i in 2..=n {
+        result *= i as f64;
+    }
+    result
+}
```
[the remark this answers is not in the document]
Comment (BABTUNA) [reply src/daft-functions/src/numeric/factorial.rs]: oh shoot that's hella true

File: tests/recordbatch/numeric/test_numeric.py [hunk 2]
```diff
@@ -843,3 +843,80 @@ def test_table_shift_right_syntactic_sugar() -> None:
     table = MicroPartition.from_pydict({"a": [1, 2, 4]})
     shift_table = table.eval_expression_list([col("a").shift_right(1)])
     assert [1 >> 1, 2 >> 1, 4 >> 1] == shift_table.get_column_by_name("a").to_pylist()
+
+
+def test_e() -> None:
+    from daft.functions import e
```
[the remark this answers is not in the document]
Comment (BABTUNA) [reply tests/recordbatch/numeric/test_numeric.py:849]: nah this follows the pattern

File: src/daft-functions/src/numeric/factorial.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    fn docstring(&self) -> &'static str {
+        "Returns the factorial of a non-negative integer."
+    }
+}
+
+fn compute_factorial(n: f64) -> f64 {
+    if n.is_nan() || n < 0.0 || n.is_infinite() || n != n.floor() {
+        return f64::NAN;
```
Comment (rohitkulshreshtha) [inline src/daft-functions/src/numeric/factorial.rs]: not a blocker: FYI — the original issue was for PySpark parity, but that's no longer the goal. Worth noting: Spark's `factorial` returns null for negative/non-integer inputs, this returns NaN. No action needed, just documenting the divergence.

File: src/daft-functions/src/numeric/factorial.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        Ok(Field::new(field.name, DataType::Float64))
+    }
+
+    fn docstring(&self) -> &'static str {
+        "Returns the factorial of a non-negative integer."
+    }
+}
+
+fn compute_factorial(n: f64) -> f64 {
+    if n.is_nan() || n < 0.0 || n.is_infinite() || n != n.floor() {
```
Comment (rohitkulshreshtha) [inline src/daft-functions/src/numeric/factorial.rs]: nit: `n != n.floor()` → `n.fract() != 0.0` — more idiomatic for checking non-integer

File: src/daft-functions/src/numeric/factorial.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        let UnaryArg { input } = inputs.try_into()?;
+        let field = input.to_field(schema)?;
+        if !field.dtype.is_numeric() {
+            return Err(DaftError::TypeError(format!(
+                "Expected input to factorial to be numeric, got {}",
+                field.dtype
+            )));
+        }
+        Ok(Field::new(field.name, DataType::Float64))
+    }
```
Comment (rohitkulshreshtha) [inline src/daft-functions/src/numeric/factorial.rs:45]: suggestion: PySpark's `factorial` returns `LongType` (i64) and caps at 20. This returns Float64 and caps at 170 — wider range, but loses integer precision starting at 19! since f64 can't exactly represent integers above 2^53. Since the issue is PySpark parity and both domain and range are integers, would Int64 (capping at 20) be more appropriate? Also, factorial is only defined on non-negative integers — accepting float input and returning NaN for non-integers (e.g., `factorial(3.5)`) is a runtime check for what should be a type error. If someone wants factorial for real numbers, that's the gamma function.

**Comment by rohitkulshreshtha** [comment]
Thanks for knocking out 4 functions from #3793! A couple of thoughts on factorial, the rest looks good.

File: src/daft-functions/src/numeric/factorial.rs (hunk 3 above)
Comment (BABTUNA) [reply src/daft-functions/src/numeric/factorial.rs]: ahhh makes sense, lemme change it

File: src/daft-functions/src/numeric/factorial.rs (hunk 4 above)
Comment (BABTUNA) [reply src/daft-functions/src/numeric/factorial.rs]: sounds good!

File: src/daft-functions/src/numeric/factorial.rs (hunk 5 above)
Comment (BABTUNA) [reply src/daft-functions/src/numeric/factorial.rs:45]: holly math theory, haha this just gave me discrete math flashbacks

Comment (BABTUNA) [reply src/daft-functions/src/numeric/factorial.rs:45]: ok I think for this one it makes sense to cap at 20

File: src/daft-functions/src/numeric/factorial.rs (hunk 4 above)
Comment (rohitkulshreshtha) [reply src/daft-functions/src/numeric/factorial.rs]: Ignore this suggestion - since we plan to make the domain be the set of integers as well.

**Comment by rohitkulshreshtha** [comment]
suggestion: now that the return type is Int64, consider going i64 through and through — restrict the input to integer types in `get_return_field` (matches PySpark, which only accepts `IntegralType`), and drop the f64 intermediate:

```rust
fn compute_factorial(n: i64) -> Option<i64> {
    if n < 0 || n > 20 {
        return None;
    }
    let mut result: i64 = 1;
    for i in 2..=n {
        result *= i;
    }
    Some(result)
}

fn factorial_impl(s: Series) -> DaftResult<Series> {
    let casted = s.cast(&DataType::Int64)?;
    let i64_arr = casted.i64().unwrap();
    // ...
}
```

Drops the NaN/infinity/fract checks (unreachable for integer input) and the i64 → f64 → u64 round-trip. Trade-off: `factorial(3.0_f64)` fails at type check, but that matches PySpark.

File: src/daft-functions/src/numeric/factorial.rs (hunk 5 above)
Comment (BABTUNA) [reply src/daft-functions/src/numeric/factorial.rs:45]: also change the input  to only real numbers

**Comment by BABTUNA** [comment]
alright I think that fixes everything 
image

**Review by rohitkulshreshtha — APPROVED** [review]
LGTM — the i64-through-and-through rewrite is clean. The range check and test coverage look solid. Thanks!
